### ISSUE TITLE
webchat.js - doesGlobalFileReaderExist causes EvalError when Content Security Policy is set

### ISSUE THREAD

**Issue by silverbulletgt** [microsoft/BotFramework-WebChat#3743]
## Screenshots
Browser Console showing EvalError:
![image](https://user-images.githubusercontent.com/76121744/108755329-1f033900-7515-11eb-85b4-1ced4e3d5803.png)
Line which the error occurs on:
![image](https://user-images.githubusercontent.com/76121744/108755456-42c67f00-7515-11eb-99ab-73bfb42430a4.png)

## Version

WebChat via CDN: https://cdn.botframework.com/botframework-webchat/latest/webchat.js
Version: 4.11.0

## Describe the bug

When using webchat.js from the CDN & Content Security Policies are set a browser console error is triggered & the control does not load successfully.

## Steps to reproduce

Containing Folder 
[WebChatEvalError.zip](https://github.com/microsoft/BotFramework-WebChat/files/6024547/WebChatEvalError.zip)

1. Copy files in zip file to directory: C:\inetpub\wwwroot
2. Open IIS & convert the WebChatEvalError directory to an application using an application pool which uses .NET 4
3. Open a browser & navigate to: http://localhost/WebChatEvalError/webchat.html
4. Open the browser console
5. See error

## Expected behavior

WebChat control loaded onto a page using the CDN should work when Content Security Policies are set

## Additional context

I am able to get around this issue by changing the Content Security Policy to allow unsafe-eval though this is not something I am allowed to do within my organizations policies.

[Bug]

**Comment by compulim** [comment]
@stevengum is this one fixed in latest version of DLJS + BFStreaming?

**Comment by compulim** [comment]
Some investigations with @Zerryth:

- This PR should have removed the code, https://github.com/microsoft/botbuilder-js/pull/2752
  - Shipped as 4.10.0
- [DLJS 0.13.1 is using BFStreaming 4.10.3](https://github.com/microsoft/BotFramework-DirectLineJS/blob/master/CHANGELOG.md#0131---2020-09-09)
- [Web Chat 4.10.1 is using DLJS 0.13.1](https://github.com/microsoft/BotFramework-WebChat/blob/master/CHANGELOG.md#changed-2)

So, Web Chat >= 4.10.1 should have it fixed.

We still need to extract the ZIP file from @silverbulletgt and see how it repro.

**Comment by carlosscastro** [comment]
@compulim and @Zerryth, where are we on this? Could you send an update please?

**Comment by peterinnesmsft** [comment]
@compulim any updates on this issue?

**Comment by tdurnford** [comment]
@compulim any updates on this issue?

**Comment by tdurnford** [comment]
@stevkan or someone on the support team can you verify that is resolved in latest version of Web Chat?

**Comment by stevkan** [comment]
@tdurnford, I am able to repro the issue, using the current version of Web Chat, v4.13.0.

**Comment by corinagum** [comment]
Thanks for the repro. Added it to dev discussions

**Comment by Zerryth** [comment]
@stevengum @compulim my investigations so far, as I'm OOF and handing this over:
___
So I was not able to successfully reproduce the issue, however I also wasn't able to repro it the exact way I wanted either--something's wrong with Bot Services :(

But that being said, I was able to successfully have a bot that already exists (the streaming bot) successfully chat
![image](https://user-images.githubusercontent.com/35248895/118310618-b7b9ad00-b4a3-11eb-8878-c2fc1ea94f2f.png)

___
Why I can't reproduce it with a fresh, new bot -- Bot Services issues abound

There are some  weird issue happening with Bot Services--it appears to be down, therefore I cannot access Channels > Direct Line > get secret for any new bots. Nor could I provision any new bots :(

*trying to get secret from BCR > Channels* 
![image](https://user-images.githubusercontent.com/35248895/118310286-411caf80-b4a3-11eb-97ec-c53645c085a0.png)
* "dev.botframework.com took too long to response"

*trying to create a new bot* -- need any DL secret
![image](https://user-images.githubusercontent.com/35248895/118309981-d8cdce00-b4a2-11eb-9aa3-5dbf5efddd86.png)
* Can't reach Bot Services

So I'm working with the one secret I have stored, and that's with (the streaming bot). 

___

My Repro so far.
1. Follow @silverbulletgt's instructions for 1 - 3.
![image](https://user-images.githubusercontent.com/35248895/118310879-02d3c000-b4a4-11eb-8977-3a6209079d66.png)

2. Edit OG's given html file to fetch a "real" token instead of passing "TestToken"

```html
<!DOCTYPE html>
<html>
   <body>
      <div id="webchat" role="main"></div>
      <script src="https://cdn.botframework.com/botframework-webchat/latest/webchat.js"></script>
      <script>

         // Set style options.
         const styleOptions = {
            botAvatarInitials: 'BF',
            userAvatarInitials: 'WC'
         };
		 
		 //trying to apply fix for issue: https://github.com/microsoft/BotFramework-WebChat/issues/3393
		 //which does not resolve this issue
		 const styleNonce = "testNonce";

       (async function() {
        const res = await fetch('https://directline.botframework.com/v3/directline/tokens/generate', { method: 'POST', headers:{'Authorization':'Bearer IVOl9nxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxs6i_FtiM'} });
        const { token } = await res.json();
        
         window.WebChat.renderWebChat(
            {
               directLine: window.WebChat.createDirectLine({
                  token: token
               }),
               userID: 'YOUR_USER_ID',
               username: 'Web Chat User',
               locale: 'en-US',
               styleOptions,
               styleNonce
            },
            document.getElementById('webchat')
         );

         document.querySelector('#webchat > *').focus();
         })().catch(err => console.error(err));
      </script>
   </body>
</html>
```

3. Navigate to inetpub/wwwroot/webchat.html => success
![image](https://user-images.githubusercontent.com/35248895/118310929-141ccc80-b4a4-11eb-835f-ba26935f6ede.png)

**Comment by Zerryth** [comment]
Also note: for anyone looking at the HTML, please be sure to add a token server in your production code, rather than putting the Direct Line secret into the HTML like I had done for testing

**Comment by stevengum** [comment]
@compulim I'm working with @stevkan and @Zerryth to unpack this.

They were unable to repro the `unsafe-eval` CSP violation using latest and 4.13.0 via the CDN. @stevekan is building WC locally and attempting a another repro.

**Comment by stevengum** [comment]
@silverbulletgt can you bump your version of Web Chat to 4.12.0 or greater? I've checked that version on the CDN and the problematic unsafe-eval code from botframework-streaming was not found there.

@compulim when inspecting [4.11.0](https://cdn.botframework.com/botframework-webchat/4.11.0/webchat.js) and [4.10.1](https://cdn.botframework.com/botframework-webchat/4.10.1/webchat.js) on the CDN, I was able to find the problematic `doesGlobalFileReaderExist` function that was removed from `botframework-streaming@4.10.0`.  

As @Zerryth [found earlier](https://github.com/microsoft/BotFramework-WebChat/issues/3743#issuecomment-783682124), everything looks correct, but it's unclear as to why 4.10.1 and 4.11.0 of Web Chat still use the unsafe eval code.

> Some investigations with @Zerryth:
> 
> * This PR should have removed the code, [microsoft/botbuilder-js#2752](https://github.com/microsoft/botbuilder-js/pull/2752)
>   
>   * Shipped as 4.10.0
> * [DLJS 0.13.1 is using BFStreaming 4.10.3](https://github.com/microsoft/BotFramework-DirectLineJS/blob/master/CHANGELOG.md#0131---2020-09-09)
> * [Web Chat 4.10.1 is using DLJS 0.13.1](https://github.com/microsoft/BotFramework-WebChat/blob/master/CHANGELOG.md#changed-2)
> 
> So, Web Chat >= 4.10.1 should have it fixed.
> 
> We still need to extract the ZIP file from @silverbulletgt and see how it repro.

**Comment by stevengum** [comment]
@silverbulletgt, I’m closing this issue for now as the CSP violation via the unsafe-eval error has been verified fixed in later versions of Web Chat (starting with 4.12.0).

Please bump the version of Web Chat being used in your front end to pick up this fix. If you still run into the same CSP violations with a version equal to or greater than 4.12.0, please file a new issue.
